extensions/binding/org.eclipse.smarthome.binding.tradfri/src/main/java/org/eclipse/smarthome/binding/tradfri/handler/TradfriGatewayHandler.java

```diff
@@ -77,20 +85,53 @@ public void handleCommand(ChannelUID channelUID, Command command) {
-        if (configuration.host == null || configuration.host.isEmpty()) {
+
+        if (isNullOrEmpty(configuration.host)) {
             updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
                     "Host must be specified in the configuration!");
             return;
         }
-        if (configuration.code == null || configuration.code.isEmpty()) {
-            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
-                    "Security code must be provided in the configuration!");
-            return;
+        if (isNullOrEmpty(configuration.code)) {
+            if (isNullOrEmpty(configuration.identity) || isNullOrEmpty(configuration.preSharedKey)) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                        "Either security code or identity and pre-shared key must be provided in the configuration!");
+                return;
+            }
+        } else {
+            if (isOldFirmware()) {
+                /*
+                 * older firmware - fall back to authentication with security code
+                 * in this case the Thing configuration will not be persisted
+                 */
+                logger.warn("Gateway with old firmware - please consider upgrading to the latest version.");
+                configuration.identity = "";
+                configuration.preSharedKey = configuration.code;
+            } else {
+                configuration.identity = UUID.randomUUID().toString().replace("-", "");
+                configuration.preSharedKey = authenticateOnGateway(configuration);
+
+                if (isNullOrEmpty(configuration.preSharedKey)) {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Pre-shared key was not obtain successfully");
+                    return;
+                } else {
+                    logger.info(
+                            "Successfully obtained pre-shared key for identity {}. Storing it in thing configuration and discarding the securiy code.",
+                            configuration.identity);
+                    logger.debug("Pair<identity, key> => <{}, {}>", configuration.identity, configuration.preSharedKey);
+
+                    configuration.code = null;
+
+                    Configuration editedConfig = editConfiguration();
+                    editedConfig.put(TradfriBindingConstants.GATEWAY_CONFIG_CODE, configuration.code);
+                    editedConfig.put(TradfriBindingConstants.GATEWAY_CONFIG_IDENTITY, configuration.identity);
+                    editedConfig.put(TradfriBindingConstants.GATEWAY_CONFIG_PRE_SHARED_KEY, configuration.preSharedKey);
+                    updateConfiguration(editedConfig);
+                }
+            }
         }
 
         this.gatewayURI = "coaps://" + configuration.host + ":" + configuration.port + "/" + DEVICES;
```

Comment:
You should not continue to work on the outdated `configuration` instance, but rather retrieve a fresh one after you have called `updateConfiguration`.

Reply:
how about using the variables which I will introduce and which will be used for updating the configuration ?

Reply:
You do not know whether `updateConfiguration()` might do any magic and tweak your config in some way. So: No.

Reply:
I would never expect this from ```updateConfiguration()``` :)
but OK - will retrieve the config once again after it just been set